### PR TITLE
fix: `FrameworkTemplate` pool with `TextBox` with TwoWay binding

### LINKED ISSUE #13125: Template-recycled `TextBox` can reset `TwoWay` data-binding source
## For which Platform:

- [ ] WebAssembly
- [ ] Android
- [ ] iOS
- [ ] macOS (AppKit)
- [ ] Mac Catalyst
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Linux Framebuffer
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): closes #13125

## PR Type

What kind of change does this PR introduce?

- Bugfix

### <samp>🤖 Generated by Copilot at c964b0a</samp>

This pull request adds new unit tests and fixes some issues related to the template pooling and recycling feature for the `TextBox`, `ToggleButton`, and `ToggleSwitch` controls. It also enables template pooling for the samples app and fixes a bug related to the window size and the layout. It introduces new classes and files for the unit tests, such as `EditorTestPage`, `EditorViewModel`, and `EditorTemplateSelector`, and modifies some existing files, such as `BindingExpression.cs` and `ToggleButton.cs`.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 1]
```diff
@@ -346,7 +351,9 @@ internal static void PropagateOnTemplateReused(object instance)
 			// If DataContext is not null, it means it has been explicitly set (not inherited). Resetting the view could push an invalid value through 2-way binding in this case.
 			if (instance is IFrameworkTemplatePoolAware templateAwareElement && (instance as IFrameworkElement)!.DataContext == null)
 			{
+				IsRecycling = true;
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:354]: Why is it a static property?

I would imagine this to be a property in `IFrameworkTemplatePoolAware` and set it on the specific instance whose template is being recycled

**Review by Youssef1313 — APPROVED** [review]
🎉

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:354]: The reason is that it is the `BindingExpression` that is then checking for whether a recycling is in progress, which makes it a bit tricky to get the owner view to check if it being recycled. Currently, the only thing the template aware types do is they change a single property, and it all needs to happen on the UI thread, so this seems quite safe. If I did it instance based, all two way bindings would then first need to get the owner somehow (by accessing a weakref probably) and then checking if it is in fact instace of `IFrameworkTemplatePoolAware` and then check the property. So this simpler approach seems to me better perf-wise

**Review by ajpinedam — APPROVED** [review]

**Review by Youssef1313 — APPROVED** [review]

**Comment by MartinZikmund** [comment]
@mergifyio backport legacy/4x

File: src/Uno.UI/DataBinding/BindingExpression.cs [hunk 2]
```diff
@@ -244,6 +249,11 @@ public void SetSourceValue(object value)
 
 			try
 			{
+				if (FrameworkTemplatePool.IsRecycling && _view.IsAlive && _view.Target is IFrameworkTemplatePoolAware)
```
Comment (jeromelaban) [inline src/Uno.UI/DataBinding/BindingExpression.cs:252]: @MartinZikmund We should be adding a small explanation on why we're doing this.

File: src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs [hunk 3]
```diff
@@ -374,8 +374,6 @@ public void UpdateContentTemplateRoot()
 			{
 				OnApplyTemplate();
 			}
-
-			SyncDataContext();
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs:378]: @MartinZikmund why was this particular statement problematic?

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs:378]: It's duplicated. The only caller of `UpdateContentTemplateRoot` already does `SyncDataContext`.

File: src/Uno.UI/DataBinding/BindingExpression.cs (hunk 2 above)
Comment (MartinZikmund) [reply src/Uno.UI/DataBinding/BindingExpression.cs:252]: Will do

File: src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs (hunk 3 above)
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs:378]: Yes, the call was duplicated, so the sync happened twice unnecessarily

**Pull request merged**